### PR TITLE
Jumplist picker and jump to row/col on existing buffers.

### PR DESCRIPTION
**Pull request by caojoshua**
First time developer for lua, neovim plugin, and open source. Please help me in this learning experience.

Partially address #715 

1. Create a `jumplist` picker
2. Jump to row/col on existing buffers (helps for jumplist and also some existing pickers ie. quickfix)

The major limitation right now is that vim `getjumplist()` does not include file text. `vim.api.nvim_buf_get_lines()` does not work because some jump targets are in unloaded buffers. I decided to only include the filename and row/col in the entries, and rely on the previewer to view contents. Open to any suggestions here.

### CODE REVIEW

File: lua/telescope/actions/set.lua
Comment (smithbm2316) [inline lua/telescope/actions/set.lua:127]: I'm not sure if this was an intentional change, or if you have an auto-formatter that maybe did this, but why did you move this `if row and col then` block outside of the `else` block above? Was this change necessary to make your new picker work?

File: lua/telescope/builtin/internal.lua
Comment (smithbm2316) [inline lua/telescope/builtin/internal.lua:931]: Just as an fyi, you should be able to use `vim.fn` here to call the `getjumplist()` function. `vim.fn` is the Neovim Lua api for calling Vimscript functions in Lua, so you could change that line to:
```lua
local jumplist = vim.fn.getjumplist()[1]
```

File: lua/telescope/builtin/internal.lua
Comment (smithbm2316) [inline lua/telescope/builtin/internal.lua:940]: I believe you should be able to replace the line in the for loop with `table.insert(sorted_jumplist, jumplist[i])` if you'd like. `table.insert` is a part of Lua's table library which inserts an item into a table. If you only give two arguments to it, it will accept the table to insert into and the item you want to insert, inserting that item at the end of the table (similar to the `append()` method in Python for example). If you give 3 args, it would take the table to insert into, the position in the table to insert the item, and then the item to insert. Just a little bit easier to read I think using `table.insert` instead :smiley:

**Review by smithbm2316 — COMMENTED** [review]
I love this! Afaik you've done a great job for this being your first experience writing lua for a neovim plugin :smile:. Thanks for the contribution and for taking an interest! I've left a few comments of minor suggestions and clarifications, but this looks pretty nice to me. I'm a pretty new maintainer here and am still learning the codebase myself, so I'll wait for feedback from one of the other maintainers before considering merging this.

You mentioned that `getjumplist()` doesn't include the file text, can you describe why this is an issue for what you built here? I noticed you can see the location in the jumplist in the preview window, what did you have in mind to implement if you were able to get access to the file text?

File: lua/telescope/actions/set.lua
Comment (caojoshua) [reply lua/telescope/actions/set.lua:127]: This is intentional. Its not necessary for the picker to work, but is necessary to set the cursor to `row, col`. Otherwise, it just jumps to where the buffer was previously left off, or line 0 of the buffer.

Regardless of whether we want this picker, this change should be merged. For example, lets say we are telescoping the quickfix list, and an open buffer has `N` entries. If we select any of those entries, the cursor wherever the buffer was left previously. With this change, the cursor would be set at the correct `row, col`.

**Comment by caojoshua** [comment]
Thanks for the review!

Let me clarify. For example, quickfix list does:
```
    return displayer {
      line_info,
      entry.text:gsub(".* | ", ""),
      filename,
    }
```
Where `entry.text` holds the actual line of code/text in the file. This works because the items returned from `vim.fn.getqflist()` have an `item` field. `vim.fn.getjumplist()` does not do the same.

Instead, I've implemented
```
    return displayer {
      line_info,
      filename,
    }
```
I would prefer it to also include the text, but there does not seem to be an easy way to do it.

**Comment by smithbm2316** [comment]
> Where `entry.text` holds the actual line of code/text in the file. This works because the items returned from `vim.fn.getqflist()` have an `item` field. `vim.fn.getjumplist()` does not do the same.
> 
> I would prefer it to also include the text, but there does not seem to be an easy way to do it.

Oh I see! Unfortunately I don't know of any other vimscript functions that might make that possible. I'm relatively new to Lua and Neovim plugin development myself, so I'm still learning the APIs. With the new changes this LGTM, as I mentioned before I'll wait for another maintainer to give this the green light before merging.

**Comment by smithbm2316** [comment]
@tami5 @elianiva any chance one of you could take a peek at this PR? If you think it looks good I can merge, I'm still not too familiar with the `make_entry.lua` section of the codebase yet.

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:941]: Can't we merge these two iterations

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:944]: wrong indent

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [inline lua/telescope/builtin/internal.lua:949]: should be `previewer = conf.qflist_previewer(opts),`

File: lua/telescope/actions/set.lua
Comment (Conni2461) [reply lua/telescope/actions/set.lua:127]: I think this fixes: https://github.com/nvim-telescope/telescope.nvim/issues/794

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [reply lua/telescope/builtin/internal.lua:949]: Oh i think its also wrong for `tagstack`, can you fix that as well

File: lua/telescope/builtin/internal.lua
Comment (Conni2461) [reply lua/telescope/builtin/internal.lua:944]: Indent in `tagstack` is also wrong.

File: lua/telescope/actions/set.lua
Comment (caojoshua) [reply lua/telescope/actions/set.lua:127]: oh sweet didn't know there was already an issue for this. Yes I was able to reproduce the issue and fix it.

**Review by tami5 — APPROVED** [review]

**Review by smithbm2316 — APPROVED** [review]

**Comment by smithbm2316** [comment]
@caojoshua thanks for all the hard work on this! I'll merge this once #821 is resolved (that is what is causing the two tests to fail atm)

**Review by Conni2461 — APPROVED** [review]
LGTM.

Can we get a mention of that picker in the README?

**Comment by caojoshua** [comment]
Thanks everyone. Left another commit for adding things to README

File: lua/telescope/make_entry.lua
Comment (Conni2461) [inline lua/telescope/make_entry.lua:1165]: Just tested it. I am getting: `Invalid buffer id: 1` here if i had startify open before.
So: `nvim` -> `:Telescope find_files` -> `open file` -> `:Telescope jumplist` (error)
If i do: `nvim file` -> `:Telescope jumplist` works great

We should do
```lua
if not vim.api.nvim_buf_is_valid(entry.bufnr) then
  return
end
```
before this line. Thoughts?

Comment (Conni2461) [reply lua/telescope/make_entry.lua:1165]: Otherwise works great :) thanks

Comment (caojoshua) [reply lua/telescope/make_entry.lua:1169]: Added this to the commit.

I personally don't use startify, and haven't been able to reproduce a scenario where this makes/breaks anything. I think its OK, unless there is some other sort of verification we can do.

**Comment by Conni2461** [comment]
Thanks all the work. Let's get this merged :)

**Pull request merged**
